File: app/src/main/java/dev/gavenda/yuuka/sync/Outbox.kt
package dev.gavenda.yuuka.sync

import dev.gavenda.yuuka.data.local.dao.OutboxDao
import dev.gavenda.yuuka.data.local.entity.OutboxEntity
import dev.gavenda.yuuka.data.remote.ApiError
import dev.gavenda.yuuka.data.remote.dto.SyncBatchRequest
import dev.gavenda.yuuka.data.remote.YuukaApi
import dev.gavenda.yuuka.data.remote.apiCall
import dev.gavenda.yuuka.data.remote.apiJson
import dev.gavenda.yuuka.data.remote.dto.SyncOperationDto
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.JsonElement
import java.time.Instant

/**
 * Local-first writes, and the batch that drains them.
 *
 * Every mutation in the app goes through [enqueue], and [enqueue] never waits
 * for the network. It writes the change to the outbox and returns; the caller
 * has already written the row it drew into Room, so the screen — which observes
 * Room, not the network — updates at once, with or without a connection.
 *
 * The queue drains separately: right away when [enqueue] is called, when the
 * network comes back ([NetworkMonitor]), and from [SyncWorker] under a
 * connectivity constraint so a queue built offline is still sent if the app is
 * never reopened. It drains as one `POST /api/sync/batch`, in the order the
 * user made the changes, which is what makes "spend from the account I just
 * made" work after an hour with no signal.
 *
 * Three things follow from writing first and sending later:
 *
 * - **The row written locally is a guess, and the server's is the truth.** Once
 *   a batch lands, [synced] names the slices it touched and the repositories
 *   refetch them, so any figure worked out approximately here — a balance, a
 *   running total — is corrected by the only party that can be sure.
 * - **A rejection arrives late.** A name the server considers a duplicate is
 *   only known to be one when the batch lands, which may be a day later. That
 *   is surfaced through [rejections] and the optimistic row is dropped by the
 *   refresh, rather than pretending the save is still in question.
 * - **Sending twice is better than losing one.** A row leaves the queue only
 *   once the server has accounted for it, so a flush that never got an answer
 *   sends again. The API is built for that.
 */
class Outbox(
    private val api: YuukaApi,
    private val dao: OutboxDao,
) {
    /** Given up on after this many failed sends, so one impossible change cannot block the queue forever. */
    private val maxAttempts = 8

    /** A batch this size already covers a long time offline; the rest goes in the next one. */
    private val batchSize = 200

    private val flushLock = Mutex()

    /** Outlives any screen, so a save made just before leaving one is still sent. */
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _syncing = MutableStateFlow(false)
    private val _synced = MutableSharedFlow<List<String>>(extraBufferCapacity = 8)
    private val _rejections = MutableSharedFlow<String>(extraBufferCapacity = 8)

    /** How many changes have not reached the API. Zero is "everything is saved". */
    val unsentCount: Flow<Int> = dao.observeCount()

    /** Whether a batch is in flight, for a screen to say "syncing". */
    val syncing: StateFlow<Boolean> = _syncing.asStateFlow()

    /** Emits the slices a landed batch touched, so the repositories holding them refetch. */
    val synced: SharedFlow<List<String>> = _synced.asSharedFlow()

    /** Emits what the server refused, for a screen to surface — the user is told late, but told. */
    val rejections: SharedFlow<String> = _rejections.asSharedFlow()

    /**
     * Queues a mutation. Returns as soon as it is written down.
     *
     * `at` is stamped here, when the user made the change, rather than when it
     * is sent: that is what the server judges a stale edit against, and a change
     * made on a plane should lose to one made on the ground an hour later, not
     * win by arriving second.
     *
     * [entity] and [rowId] are what the server checks that against, so an edit
     * or a delete should name the row it touches. A create names nothing —
     * there is no row for it to be stale against.
     */
    suspend fun enqueue(
        method: String,
        path: String,
        body: JsonElement? = null,
        entity: String? = null,
        rowId: String? = null,
    ) {
        dao.add(
            OutboxEntity(
                opId = java.util.UUID.randomUUID().toString(),
                method = method,
                path = path,
                at = Instant.now().toString(),
                entity = entity,
                rowId = rowId,
                body = body?.let { apiJson.encodeToString(JsonElement.serializer(), it) },
                slices = Slices.forPath(path).joinToString(","),
            ),
        )

        // Sent in the background: the caller is told the change is saved the moment it is written down.
        scope.launch { runCatching { flush() } }
    }

    /**
     * Sends everything waiting, as one batch.
     *
     * Only one runs at a time — a second call while one is in flight waits for
     * it rather than sending the same rows again. Nothing here decides on
     * whether the device thinks it is online: a flush is always attempted, and
     * an unreachable API simply leaves the queue as it was.
     *
     * Returns false when there is still work that did not get through, which is
     * what [SyncWorker] retries on.
     */
    suspend fun flush(): Boolean = flushLock.withLock {
        val entries = dao.oldest(batchSize)
        if (entries.isEmpty()) return@withLock true

        _syncing.value = true
        try {
            val response = try {
                apiCall {
                    api.syncBatch(
                        SyncBatchRequest(
                            entries.map { entry ->
                                SyncOperationDto(
                                    opId = entry.opId,
                                    method = entry.method,
                                    path = entry.path,
                                    at = entry.at,
                                    entity = entry.entity,
                                    id = entry.rowId,
                                    body = entry.body?.let { apiJson.parseToJsonElement(it) },
                                )
                            },
                        ),
                    )
                }
            } catch (error: ApiError) {
                // Still offline, or the session expired. The queue is untouched
                // and will be sent again; the only thing recorded is that the
                // attempt happened, so a row the server will never accept is
                // eventually given up on.
                dao.recordFailure(entries.map { it.seq }, error.message)
                return@withLock false
            }

            val bySeq = entries.associateBy { it.opId }
            val settled = mutableListOf<Long>()
            val retry = mutableListOf<Long>()
            val touched = linkedSetOf<String>()
            val refused = mutableListOf<String>()

            for (result in response.results) {
                val entry = bySeq[result.opId] ?: continue

                when {
                    result.status == "applied" || result.status == "stale" -> {
                        settled += entry.seq
                        touched += entry.slices.split(',').filter { it.isNotBlank() }
                    }
                    // The server answered, and said no. Retrying a 4xx would only
                    // be told the same thing, so it is given up on at once — the
                    // attempt count is for a batch that never got through.
                    result.code in 400..499 || entry.attempts + 1 >= maxAttempts -> {
                        settled += entry.seq
                        touched += entry.slices.split(',').filter { it.isNotBlank() }
                        result.error?.let { refused += it }
                    }
                    else -> retry += entry.seq
                }
            }

            if (settled.isNotEmpty()) dao.remove(settled)
            if (retry.isNotEmpty()) dao.recordFailure(retry, "Retrying.")

            // The provisional rows are replaced by what the server actually has,
            // which is also what takes a rejected change back off the screen.
            if (touched.isNotEmpty()) _synced.tryEmit(touched.toList())
            for (message in refused) _rejections.tryEmit(message)

            retry.isEmpty() && dao.count() == 0
        } finally {
            _syncing.value = false
        }
    }

    /** Forgets the unsent work. Signing out — it was written against a session that is over. */
    suspend fun discard() = dao.clear()
}

File: webapp/src/lib/outbox.ts
/**
 * The queue of changes that have not reached the API yet.
 *
 * This is the one thing in the browser that is **not** a cache. Everything in
 * `cache.ts` can be thrown away and refetched; an entry here is a change the
 * user made that exists nowhere else yet, so losing it loses their work. That
 * is why it is in IndexedDB rather than beside the cache in `localStorage`:
 * the cache drops its oldest entries when the quota is tight, and the queue
 * must never be what gets dropped.
 *
 * Two properties the rest of the app depends on:
 *
 * - **Order is preserved.** Entries are keyed by an auto-incrementing sequence
 *   and always read in that order. Creating an account and then spending from
 *   it is two entries, and the second only works after the first.
 * - **An entry is removed only once the server has accounted for it.** A flush
 *   that never got an answer leaves the queue as it was, and sends again. The
 *   API is built to be sent the same batch twice, so a duplicate send is
 *   cheaper than a lost change.
 *
 * The queue holds one person's unsent work, so it is emptied when someone else
 * signs in — the same rule the cache follows, for the same reason.
 */

import type { ChangeSlice } from './slices';

const DATABASE = 'yuuka-outbox';
const STORE = 'operations';
const VERSION = 1;

/** One queued call, in the shape `POST /api/sync/batch` takes. */
export interface OutboxEntry {
	/** Assigned by IndexedDB; the queue's order and the key to remove it by. */
	seq?: number;
	/** The client's own name for this operation, echoed back in the batch's results. */
	opId: string;
	method: 'POST' | 'PATCH' | 'PUT' | 'DELETE';
	/** An `/api/...` path, exactly as the same call would look online. */
	path: string;
	/** When the user made the change, not when it is sent — this is what a stale edit is judged against. */
	at: string;
	/** What it touches, for the server's staleness check. */
	entity?: string;
	id?: string;
	body?: unknown;
	/** Which slices to refresh once this lands or fails, so the screens catch up. */
	slices: ChangeSlice[];
	/** How many times sending this has been attempted, so a permanently broken entry can be given up on. */
	attempts: number;
	/** Why the last attempt failed, kept so the interface can say what went wrong. */
	lastError?: string;
}

let opening: Promise<IDBDatabase | null> | null = null;

/**
 * Opens the database, once. A browser that refuses IndexedDB — a locked-down
 * private window — resolves to null, and every function here then behaves as
 * if the queue were empty: writes go straight to the API and fail when
 * offline, which is how the app behaved before there was a queue at all.
 */
function open(): Promise<IDBDatabase | null> {
	if (opening) return opening;

	opening = new Promise<IDBDatabase | null>((resolve) => {
		let request: IDBOpenDBRequest;
		try {
			request = indexedDB.open(DATABASE, VERSION);
		} catch {
			resolve(null);
			return;
		}

		request.onupgradeneeded = () => {
			const db = request.result;
			if (!db.objectStoreNames.contains(STORE)) db.createObjectStore(STORE, { keyPath: 'seq', autoIncrement: true });
		};
		request.onsuccess = () => resolve(request.result);
		request.onerror = () => resolve(null);
		request.onblocked = () => resolve(null);
	});

	return opening;
}

function run<T>(mode: IDBTransactionMode, work: (store: IDBObjectStore) => IDBRequest<T>): Promise<T | null> {
	return open().then(
		(db) =>
			new Promise<T | null>((resolve) => {
				if (!db) {
					resolve(null);
					return;
				}

				try {
					const transaction = db.transaction(STORE, mode);
					const request = work(transaction.objectStore(STORE));
					request.onsuccess = () => resolve(request.result);
					request.onerror = () => resolve(null);
					transaction.onabort = () => resolve(null);
				} catch {
					resolve(null);
				}
			}),
	);
}

/** Adds an entry to the back of the queue and returns it with the sequence it was given. */
export async function enqueue(entry: Omit<OutboxEntry, 'seq' | 'attempts'>): Promise<OutboxEntry | null> {
	const record: OutboxEntry = { ...entry, attempts: 0 };
	const seq = await run<IDBValidKey>('readwrite', (store) => store.add(record) as IDBRequest<IDBValidKey>);
	if (seq === null) return null;

	return { ...record, seq: Number(seq) };
}

/** Everything waiting, oldest first. */
export async function pending(): Promise<OutboxEntry[]> {
	return (await run<OutboxEntry[]>('readonly', (store) => store.getAll() as IDBRequest<OutboxEntry[]>)) ?? [];
}

/** How many changes have not reached the API. What the interface counts when it says "3 unsynced". */
export async function pendingCount(): Promise<number> {
	return (await run<number>('readonly', (store) => store.count() as IDBRequest<number>)) ?? 0;
}

/** Drops entries the server has accounted for — applied, dropped as stale, or given up on. */
export async function remove(sequences: number[]): Promise<void> {
	if (sequences.length === 0) return;

	const db = await open();
	if (!db) return;

	await new Promise<void>((resolve) => {
		try {
			const transaction = db.transaction(STORE, 'readwrite');
			const store = transaction.objectStore(STORE);
			for (const seq of sequences) store.delete(seq);
			transaction.oncomplete = () => resolve();
			transaction.onerror = () => resolve();
			transaction.onabort = () => resolve();
		} catch {
			resolve();
		}
	});
}

/** Records that sending these failed, so a permanently broken entry can eventually be given up on. */
export async function recordFailure(entries: { seq: number; error: string }[]): Promise<void> {
	if (entries.length === 0) return;

	const db = await open();
	if (!db) return;

	await new Promise<void>((resolve) => {
		try {
			const transaction = db.transaction(STORE, 'readwrite');
			const store = transaction.objectStore(STORE);
			for (const { seq, error } of entries) {
				const read = store.get(seq);
				read.onsuccess = () => {
					const record = read.result as OutboxEntry | undefined;
					if (!record) return;
					store.put({ ...record, attempts: record.attempts + 1, lastError: error });
				};
			}
			transaction.oncomplete = () => resolve();
			transaction.onerror = () => resolve();
			transaction.onabort = () => resolve();
		} catch {
			resolve();
		}
	});
}

/** Empties the queue. Signing out, or someone else signing in — the unsent work is not theirs. */
export async function clearOutbox(): Promise<void> {
	await run('readwrite', (store) => store.clear() as IDBRequest<undefined>);
}
